### PR TITLE
aws: Support roles and profiles in AWS sources

### LINKED ISSUE #5895: aws sources: Support assuming roles directly in materialized
Users have an expectation that a correctly configured AWS `credentials` or `profile` file will cause systems to assume configured roles. Unfortunately Rusoto does not do this out of the box ([issue](https://github.com/rusoto/rusoto/issues/1758), [pr](https://github.com/rusoto/rusoto/pull/1842)) -- afaict it doesn't even provide a generic profile/credentials parser.

Adding an StsAssumedRoleProvider into our provider chain isn't a huge lift, but deciding whether we require users to specify a `role_arn` in the source constructor or do the right thing and extract the role_arn from a provided profile is important. If there are not any decent credentials parsers around then we will need to implement that as well. The above-linked PR looks reasonable to a first approximation, for this specific use case.

While S3 is in experimental mode we could do the hackier thing and require the role_arn inside the source's `WITH` clause, with an intention to never stabilize that option, which would unblock #5891 and similar use cases.

### PR DESCRIPTION
**Pull request by quodlibetor**
This enhances our support of AWS by allowing users to specify either a
`profile` or a `role_arn` as part of the Kinesis or S3 `WITH` options.

Additionally, this makes the `region` optional in S3 sources, as it may be
provided as part of default config lookup chain.

### Motivation

* This PR adds a known-desirable feature: resolves #5895

### Tips for reviewers

The files changed metric is misleading, the overwhelming majority of that is piping a CLI flag through to SQL. I put that flag into a separate commit, if you review this commit by commit it should be pretty trivial.

### Checklist

Both of these are still TODO

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note)L https://github.com/MaterializeInc/materialize/pull/10216

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/9894)

### CODE REVIEW

**Comment by quodlibetor** [comment]
@benesch CCing you on this because the first two commits modify `mzcompose/__init__.py`

File: src/dataflow-types/src/types.rs [hunk 1]
```diff
@@ -1307,42 +1313,115 @@ pub mod sources {
...
+                    let mut wc =
+                        aws_config::default_provider::region::DefaultRegionChain::builder();
+                    if let AwsCredentials::Profile { profile_name } = &self.credentials {
+                        wc = wc.profile_name(profile_name);
+                    }
+                    // TODO: this doesn't work for profiles which do not provide their own region
+                    // but instead specify a source_profile that does have a region
+                    RegionProviderChain::first_try(wc.build()).or_default_provider()
+                }
+            };
```
Comment (benesch) [inline src/dataflow-types/src/types.rs:1373]: There's some nested provider chains going on here that seem weird. I think it's simpler to just return `Box<dyn ProvideRegion>` here to return the two types directly:

```rust
            let make_region_provider = || match &self.region {
                Some(region) => Box::new(Region::new(region.clone())) as Box<dyn ProvideRegion>,
                _ => {
                    let mut rc = DefaultRegionChain::builder();
                    if let AwsCredentials::Profile { profile_name } = &self.credentials {
                        rc = rc.profile_name(profile_name);
                    }
                    // TODO: this doesn't work for profiles which do not provide their own region
                    // but instead specify a source_profile that does have a region
                    Box::new(rc.build())
                }
            };
```

Comment (benesch) [reply src/dataflow-types/src/types.rs:1373]: Actually, I looked into what the SDK does internally, and you might as well just load the region here. It's much simpler that way:

```rust
            let region = match &self.region {
                Some(region) => Some(Region::new(region.clone())),
                _ => {
                    let mut rc = DefaultRegionChain::builder();
                    if let AwsCredentials::Profile { profile_name } = &self.credentials {
                        rc = rc.profile_name(profile_name);
                    }
                    // TODO: this doesn't work for profiles which do not provide their own region
                    // but instead specify a source_profile that does have a region
                    rc.build().region().await
                }
            };
```

File: src/dataflow-types/src/types.rs [hunk 2]
```diff
@@ -1307,42 +1313,115 @@ pub mod sources {
...
+                Some(role)
+            } else {
+                None
+            };
+
+            loader = if let Some(role) = role {
+                loader.credentials_provider(role.build(cred_provider))
+            } else {
+                loader.credentials_provider(cred_provider)
+            };
```
Comment (benesch) [inline src/dataflow-types/src/types.rs:1423]: I think this can be combined with the last `if` statement.

File: src/dataflow-types/src/types.rs [hunk 3]
```diff
@@ -1307,42 +1313,115 @@ pub mod sources {
...
+
             let mut config = mz_aws_util::config::AwsConfig::from_loader(loader).await;
             if let Some(endpoint) = &self.endpoint {
                 config.set_endpoint(Endpoint::immutable(endpoint.0.clone()));
```
Comment (benesch) [inline src/dataflow-types/src/types.rs:1427]: So I think this whole function gets a good bit clearer if you structure it like so:

```rust
        /// Loads the AWS SDK configuration object from the environment, then
        /// applies the overrides from this object.
        pub async fn load(&self) -> mz_aws_util::config::AwsConfig {
            use aws_config::default_provider::credentials::DefaultCredentialsChain;
            use aws_config::default_provider::region::DefaultRegionChain;
            use aws_config::sts::AssumeRoleProvider;
            use aws_smithy_http::endpoint::Endpoint;
            use aws_types::credentials::SharedCredentialsProvider;
            use aws_types::region::Region;

            let region = match &self.region {
                Some(region) => Some(Region::new(region.clone())),
                _ => {
                    let mut rc = DefaultRegionChain::builder();
                    if let AwsCredentials::Profile { profile_name } = &self.credentials {
                        rc = rc.profile_name(profile_name);
                    }
                    // TODO: this doesn't work for profiles which do not provide their own region
                    // but instead specify a source_profile that does have a region
                    rc.build().region().await
                }
            };

            let mut cred_provider = match &self.credentials {
                AwsCredentials::Default => SharedCredentialsProvider::new(
                    DefaultCredentialsChain::builder()
                        .region(region.clone())
                        .build()
                        .await,
                ),
                AwsCredentials::Profile { profile_name } => SharedCredentialsProvider::new(
                    DefaultCredentialsChain::builder()
                        .profile_name(profile_name)
                        .region(region.clone())
                        .build()
                        .await,
                ),
                AwsCredentials::Static {
                    access_key_id,
                    secret_access_key,
                    session_token,
                } => SharedCredentialsProvider::new(aws_types::Credentials::from_keys(
                    access_key_id,
                    secret_access_key,
                    session_token.clone(),
                )),
            };

            if let Some(role_arn) = &self.role_arn {
                let mut role = AssumeRoleProvider::builder(role_arn).session_name("materialized");
                // This affects which region to perform STS on, not where
                // anything else happens.
                if let Some(region) = &region {
                    role = role.region(region.clone());
                }
                if let Some(external_id) = &self.external_id {
                    role = role.external_id(external_id);
                }
                cred_provider = SharedCredentialsProvider::new(role.build(cred_provider));
            }

            let loader = aws_config::from_env().region(region).credentials_provider(cred_provider);
            let mut config = mz_aws_util::config::AwsConfig::from_loader(loader).await;
            if let Some(endpoint) = &self.endpoint {
                config.set_endpoint(Endpoint::immutable(endpoint.0.clone()));
            }
            config
        }
```

Btw, I put the `use` block inside the function because most of the types (`Region`, `Endpoint`) aren't obviously AWS types, and it didn't seem right to import them into the whole massive module's scope.

File: src/dataflow-types/src/types.rs [hunk 4]
```diff
@@ -1307,42 +1313,115 @@ pub mod sources {
...
+                if let Some(region) = &self.region {
+                    RegionProviderChain::first_try(Region::new(region.clone()))
+                        .or_default_provider()
+                } else {
+                    let mut wc =
+                        aws_config::default_provider::region::DefaultRegionChain::builder();
+                    if let AwsCredentials::Profile { profile_name } = &self.credentials {
+                        wc = wc.profile_name(profile_name);
+                    }
+                    // TODO: this doesn't work for profiles which do not provide their own region
```
Comment (benesch) [inline src/dataflow-types/src/types.rs:1369]: Is this a bug? It seems like something we should file upstream.

File: src/dataflow-types/src/types.rs [hunk 5]
```diff
@@ -1307,42 +1313,115 @@ pub mod sources {
...
+                    access_key_id,
+                    secret_access_key,
+                    session_token.clone(),
+                )),
+            };
+
+            let role = if let Some(role_arn) = self.role_arn.as_ref() {
+                let mut role = AssumeRoleProvider::builder(role_arn);
+
+                // this affects which region to perform STS on, not where anything else happens
```
Comment (benesch) [inline src/dataflow-types/src/types.rs:1405]: Does it matter what region we perform STS on? Would folks ever want to perform STS in a different region than the one they want to perform other API operations on?

File: src/dataflow-types/src/types.rs [hunk 6]
```diff
@@ -1307,42 +1313,115 @@ pub mod sources {
...
 
-    /// Static AWS credentials for a source or sink.
+    /// AWS credentials for a source or sink.
     #[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
-    pub struct AwsCredentials {
-        pub access_key_id: String,
-        pub secret_access_key: String,
-        pub session_token: Option<String>,
+    pub enum AwsCredentials {
+        /// Look for credentials using the [default provider][docs] chain
```
Comment (benesch) [inline src/dataflow-types/src/types.rs:1339]: I think you want the "default credentials chain" here, and you can use Rustdoc's auto-linking facilities.

```
[default credentials chain](aws_config::default_provider::credentials::DefaultCredentialsChain)
```

File: src/dataflow-types/src/types.rs [hunk 7]
```diff
@@ -1307,42 +1313,115 @@ pub mod sources {
...
+        ///
+        /// [docs]: https://docs.rs/aws-config/latest/aws_config/default_provider/index.html
+        Default,
+        /// Load credentials using the given named profile
+        Profile { profile_name: String },
+        /// Use the credentials provided as WITH options
```
Comment (benesch) [inline src/dataflow-types/src/types.rs:1345]:
```suggestion
        /// Use the enclosed static credentials.
```

(Talking about `WITH` options at this layer doesn't seem right.)

File: src/dataflow-types/src/types.rs (hunk 4 above)
Comment (quodlibetor) [reply src/dataflow-types/src/types.rs:1369]: yeah the todo is to remind me to create a reproducible test case to see if this happens with plain boto.

File: test/aws-config/mzcompose.py [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+            c.wait_for_materialized(MZ.name)
+            c.run(TESTDRIVE.name, *td_args, "test.td")
+
+        # Stop materialized and ensure that it still works if the profile file is missing at startup
+        if args.restart:
+            c.stop(MZ.name)
+            c.rm(MZ.name)
+
+            c.up(MZ.name)
+            sleep_for_error(args.restart_sleep_duration)
```
[the remark this answers is not in the document]
Comment (quodlibetor) [reply test/aws-config/mzcompose.py:102]: yeah this used to always cause the restart tests to perma-hang (historically restarting at all caused testdrive to hang about 70% of the time anything more than 10s caused it almost always, but then it stopped reproing and I upped it to 60s and added all these flags)

This is no longer happening, though, so... probably I can just turn it down to 10s or even zero to see if it ever shows up again.

File: test/aws-config/mzcompose.py [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+
+
+def write_aws_config(c: Composition, text: str, service: str) -> None:
+    config = c.exec(service, "sh", "-c", "mkdir -p ~/.aws")
+    config = c.exec(
+        service, "sh", "-c", f"cat > ~/.aws/config <<EOF\n{text}EOF", quiet=True
+    )
+    config = c.exec(service, "sh", "-c", f"cat ~/.aws/config", capture=True)
+    if config.stdout != text:
+        raise UIError("Unable to verify that AWS config was correctly written")
```
[the remark this answers is not in the document]
Comment (quodlibetor) [reply test/aws-config/mzcompose.py:270]: yeah that definitely would work, I specifically did this to avoid ever writing these credentials (which in general need to be admin credentials) to disk.

File: test/aws-config/mzcompose.py [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+]
+TESTDRIVE = Testdrive(
+    materialized_url=f"postgres://materialize@{MZ.name}:6875",
+)
+SERVICES = [
+    MZ,
+    TESTDRIVE,
+]
+
+SEED = str(random.randrange(0, 2 ** 32))
```
[the remark this answers is not in the document]
Comment (quodlibetor) [reply test/aws-config/mzcompose.py:34]: I need access to _a_ seed here. I could extend the `Testdrive` service to expose the created seed, but that seems like it would make the Testdrive service definition different to all others.

I could also make the seeds unrelated, but this specific seed would still need to exist.

[the remark this answers is not in the document]
Comment (quodlibetor) [reply test/aws-config/mzcompose.py:34]: I can also just use `seed=int` directly.

File: src/sql/src/pure.rs [hunk 11]
```diff
@@ -829,6 +829,10 @@ async fn compile_proto(
 async fn validate_aws_credentials(config: &AwsConfig) -> Result<(), anyhow::Error> {
     let config = config.load().await;
     let sts_client = mz_aws_util::sts::client(&config);
-    let _ = sts_client.get_caller_identity().send().await?;
+    let _ = sts_client
+        .get_caller_identity()
+        .send()
+        .await
+        .map_err(|e| anyhow!("Unable to validate AWS credentials: {:#}", e))?;
```
[the remark this answers is not in the document]
Comment (quodlibetor) [reply src/sql/src/pure.rs:836]: Done.

File: src/dataflow-types/src/types.rs (hunk 5 above)
Comment (quodlibetor) [reply src/dataflow-types/src/types.rs:1405]: Right, the point of of this is that we are setting STS to use the same region as everything else. The comment might  be useless, but the exact connection between the various things in the various providers is subtle enough that it seemed worth calling out.

Comment (benesch) [reply src/dataflow-types/src/types.rs:1405]: Sorry, not complaining about the comment! Those were legitimate questions. Like, is there a region we're _supposed_ to be performing the STS AssumeRole operation on; it is sometimes distinct from the region the customer has specified in the source (e.g., you always AssumeRole in the region that is running `materialized` even if you are connecting to an S3 bucket in a different region); would the customer ever want/need to be able to configure the STS region separately.

File: test/aws-config/mzcompose.py [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+        RoleName=role_name,
+        AssumeRolePolicyDocument=json.dumps(
+            {
+                "Version": "2012-10-17",
+                "Statement": {
+                    "Effect": "Allow",
+                    "Principal": {
+                        "AWS": [
+                            current_user,
+                            "arn:aws:sts::834237029485:assumed-role/buildkite-15f2293-Role/*",
```
Comment (alex-hunt-materialize) [inline test/aws-config/mzcompose.py:166]: Are you sure you don't want to point at a specific role here?

Maybe something like (I just made this up, get your correct role arn):
```
arn:aws:iam::834237029485:role/buildkite-15f2293-Role
```
Anything assuming that role will also be able to access it.

Comment (quodlibetor) [reply test/aws-config/mzcompose.py:166]: Yeah at this point I'm just trying to figure out exactly what I'm supposed to type (that text is what I got from `aws sts get-caller-identity`, the `*` at the end is invalid), ideally this will be set in some more general way.

File: test/aws-config/mzcompose.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (quodlibetor) [reply test/aws-config/mzcompose.py]: I am no longer doing this.

File: test/aws-config/mzcompose.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (quodlibetor) [reply test/aws-config/mzcompose.py]: I have removed the exec function and no longer exec into the container.

File: src/dataflow-types/src/types.rs [hunk 13]
```diff
@@ -1369,42 +1369,111 @@ pub mod sources {
...
-        /// The optional static credentials to use.
-        ///
-        /// If unset, credentials should instead be obtained from the environment.
-        pub credentials: Option<AwsCredentials>,
+        /// AWS Credentials, or where to find them
+        pub credentials: AwsCredentials,
         /// The AWS region to use.
-        pub region: String,
+        ///
+        /// Uses the default region (looking at env vars, config files, etc) if not provideed.
```
Comment (benesch) [inline src/dataflow-types/src/types.rs:1376]:
```suggestion
        /// Uses the default region (looking at env vars, config files, etc) if not provided.
```

File: src/dataflow-types/src/types.rs [hunk 14]
```diff
@@ -1369,42 +1369,111 @@ pub mod sources {
...
+        pub region: Option<String>,
+        /// The AWS role to assume
```
Comment (benesch) [inline src/dataflow-types/src/types.rs:1378]:
```suggestion
        /// The AWS role to assume.
```

File: src/dataflow-types/src/types.rs [hunk 15]
```diff
@@ -1369,42 +1369,111 @@ pub mod sources {
...
+        },
+    }
+
+    /// A role for Materialize to assume when performing AWS API calls.
     #[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
-    pub struct AwsCredentials {
-        pub access_key_id: String,
-        pub secret_access_key: String,
-        pub session_token: Option<String>,
+    pub struct AssumeRole {
```
Comment (benesch) [inline src/dataflow-types/src/types.rs:1403]:
```suggestion
    pub struct AwsAssumeRole {
```

File: src/dataflow-types/src/types.rs [hunk 16]
```diff
@@ -1369,42 +1369,111 @@ pub mod sources {
...
+        /// The Amazon Resource Name of the Role to assume.
```
Comment (benesch) [inline src/dataflow-types/src/types.rs:1404]:
```suggestion
        /// The Amazon Resource Name of the role to assume.
```

File: src/dataflow-types/src/types.rs [hunk 17]
```diff
@@ -1369,42 +1369,111 @@ pub mod sources {
...
+        pub arn: String,
+        /// The External ID for this customer Materialize provides during role assumption.
+        ///
+        /// <https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-user_externalid.html>
+        pub external_id: Option<String>,
```
Comment (benesch) [inline src/dataflow-types/src/types.rs:1409]: I'm loathe to expand the scope of this PR further, but ideally we'd change this to be a parameter on `load`, perhaps via a type like 

```
enum AwsExternalId {
    UserSpecified(String),
    ISwearTheUserDidNotProvideOne,
}
```

to make it _really_ clear that you 100% must plumb in whatever value the user specified. Mind filing a follow-up issue for that?

File: src/materialized/src/bin/coordd.rs [hunk 18]
```diff
@@ -58,6 +58,10 @@ struct Args {
         default_value = "mzdata"
     )]
     data_directory: PathBuf,
+
+    /// An AWS External ID to be supplied to all AssumeRole operations.
```
Comment (benesch) [inline src/materialized/src/bin/coordd.rs:62]:
```suggestion
    /// An external ID to be supplied to all AWS AssumeRole operations.
```

File: src/materialized/src/bin/materialized/main.rs [hunk 19]
```diff
@@ -324,6 +324,11 @@ struct Args {
     )]
     data_directory: PathBuf,
 
+    // === AWS options. ===
+    /// An AWS External ID to be supplied to all AssumeRole operations.
```
Comment (benesch) [inline src/materialized/src/bin/materialized/main.rs:328]:
```suggestion
    /// An external ID to be supplied to all AWS AssumeRole operations.
    ///
    /// Details: <https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-user_externalid.html>
```

File: src/materialized/src/lib.rs [hunk 20]
```diff
@@ -120,6 +120,12 @@ pub struct Config {
     /// The directory in which `materialized` should store its own metadata.
     pub data_directory: PathBuf,
 
+    // === AWS options. ===
+    /// An [external id] that all aws AssumeRole operations will use, if present.
```
Comment (benesch) [inline src/materialized/src/lib.rs:124]:
```suggestion
    /// An [external ID] to be supplied to all AWS AssumeRole operations.
```

File: src/sql/src/catalog.rs [hunk 21]
```diff
@@ -176,6 +176,8 @@ pub struct CatalogConfig {
     pub safe_mode: bool,
     /// Information about this build of Materialize.
     pub build_info: &'static BuildInfo,
+    /// An External ID to provide to all AssumeRole operations
```
Comment (benesch) [inline src/sql/src/catalog.rs:179]:
```suggestion
    /// An external ID to be supplied to all AWS AssumeRole operations.
```

File: test/aws-config/test.td [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+  )
+  FORMAT TEXT;
+
+> SELECT text FROM s3_profile_allowed
+a1
+
+# Error cases
+
+# aws timeout is pretty long, so things just show up as pending but empty for awhile
+$ set-sql-timeout duration=1m
```
Comment (benesch) [inline test/aws-config/test.td:72]: Oof, a 1m timeout here for code that will very rarely get touched or break is a hard sell. Is there an easy way to get it down? If not, since nightlies now are part of the release criteria, I think we should make this a nightly test rather than something that runs on every PR.

File: test/aws-config/mzcompose.py [hunk 23]
```diff
@@ -0,0 +1,320 @@
...
+        # == Test that requires --aws-external-id has been supplied ==
+        print("+++ Test AWS External IDs")
+        c.stop("materialized")
+        c.rm("materialized")
+
+        with c.override(MZ_EID):
+            c.up("materialized")
+            c.wait_for_materialized("materialized")
+            write_aws_config(LOCAL_DIR, profile_contents)
+            c.run("testdrive-svc", *td_args, "test-externalid-present.td")
```
Comment (benesch) [inline test/aws-config/mzcompose.py:147]: Just musing, and 100% not the fault of this PR or something to address here, but using testdrive for tests like this splits the logic of the test across Python and testdrive scripts in a way that's hard to follow. I'm hoping to add the barest of SQL assertions to mzcompose soon, so that simple tests like this can be written all in Python.

**Review by benesch — COMMENTED** [review]
The code here LGTM. Thanks for slogging. I left a few stylistic nits to make the description of the external ID parameter consistent across the codebase. 

The one last thing we need before merging is an addition to the [cleanup](https://github.com/MaterializeInc/materialize/blob/main/ci/cleanup/aws.py) script to purge roles and policies whose name starts with "test". Despite the high-quality `finally` block, these things will still leak, since build agents die and builds get interrupted if you force-push at a branch. Also it occurs to me: we should use either "mztest" or "testdrive" as the prefix; "test" is a little too common for comfort, and so the cleanup script might blow away someone's legit role in the scratch account.

File: src/dataflow-types/src/types.rs (hunk 17 above)
Comment (benesch) [reply src/dataflow-types/src/types.rs:1409]: Actually, done! #10522

File: test/aws-config/test.td (hunk 22 above)
Comment (quodlibetor) [reply test/aws-config/test.td:72]: It seems like it should be settable in the profile file, I'll validate that that works.

I think it's usually closer to 10-20 seconds, it wasn't annoying enough in testing that I felt like investigating, I just chose a larger timeout than I ran into to reduce the likelihood of flakes.

**Comment by benesch** [comment]
Oh yeah, it’s not the actual testdrive timeout that I’m worried about;
setting that to some large multiple of the average observed timeout to
avoid flakes sounds great. It’s that if it’s routinely taking 30s to do
this verification, that’s a lot of CI agent time to burn on every PR.

**Review by benesch — APPROVED** [review]
Yessss!
